compiler/rustc_mir_transform/src/inline.rs

```diff
@@ -204,12 +204,21 @@ impl<'tcx> Inliner<'tcx> {
             };
 
             let arg_tuple_ty = arg_tuple.ty(&caller_body.local_decls, self.tcx);
-            let ty::Tuple(arg_tuple_tys) = arg_tuple_ty.kind() else {
-                bug!("Closure arguments are not passed as a tuple");
+
+            // We only want to unpack the tupled arguments if this instance has
+            // no spread_arg, e.g. if it's a closure def. Otherwise, both the input
+            // and outputs expect the argument to be passed like a tuple.
+            let maybe_unpacked_arg_tuple_tys = if callee_body.spread_arg.is_none() {
+                let ty::Tuple(arg_tuple_tys) = arg_tuple_ty.kind() else {
+                    bug!("Closure arguments are not passed as a tuple");
+                };
+                arg_tuple_tys.as_slice()
+            } else {
+                std::slice::from_ref(&arg_tuple_ty)
             };
```

Comment:
TBH, this whole branch about the RustCall ABI looks wrong. Should it mirror `make_call_args`, and return `Err` instead of `bug!` in error cases?

Reply:
I mean: the previous implementation already looked fishy :smile: